File: linked-list-construction/LinkedListConstruction.swift
class LinkedListConstruction {
  class Node {
    var value: Int
    var previous: Node?
    var next: Node?

    init(value: Int) {
      self.value = value
      previous = nil
      next = nil
    }
  }

  class DoublyLinkedList {
    var head: Node?
    var tail: Node?

    init() {
      head = nil
      tail = nil
    }

    func containsNodeWithValue(value: Int) -> Bool {
      var node = head
      
      while node != nil, value != node?.value {
        node = node?.next
      } 
      
      return node != nil
    }

    func remove(node: Node) {
      if node === head {
        head = head?.next
      }
      if node === tail {
        tail = tail?.previous
      }
      removeNodeBindings(node: node)
    }

    func removeNodesWithValue(value: Int) {
      var node = head
      while(node != nil) {
         var nodeToRemove = node
         node = node?.next 
         if nodeToRemove?.value == value {
           remove(node: nodeToRemove!)
         }
      }
    }

    func insertBefore(node: Node, nodeToInsert: Node) {
      if nodeToInsert === head, nodeToInsert === tail {
        return
      }
      remove(node: nodeToInsert)
      nodeToInsert.next = node
      nodeToInsert.previous = node.previous
      if node.previous == nil {
        head = nodeToInsert
      } else {
        node.previous?.next = nodeToInsert
      }
      node.previous = nodeToInsert
    }

    func insertAfter(node: Node, nodeToInsert: Node) {
      if nodeToInsert === head, nodeToInsert === tail {
        return 
      }
      remove(node: nodeToInsert)
      nodeToInsert.previous = node
      nodeToInsert.next = node.next

      if node.next == nil {
        tail = nodeToInsert
      } else {
        node.next?.previous = nodeToInsert
      }
      node.next = nodeToInsert
    }

    func setHead(node: Node) {
      if head == nil {
        head = node
        tail = node
        return
      }
      insertBefore(node: head!, nodeToInsert: node)
    }

    func setTail(node: Node) {
      if tail == nil {
        setHead(node: node)
        return
      }
      insertAfter(node: tail!, nodeToInsert: node)
    }

    func insertAtPosition(position:  Int, nodeToInsert: Node) {
      if position == 1 { setHead(node: nodeToInsert); return }
      var node = head
      var currentPosition = 1
      while node !== nil && currentPosition != position {
        currentPosition += 1
        node = node?.next
      }

      if node !== nil {
        insertBefore(node: node!, nodeToInsert: nodeToInsert)
      } else {
        setTail(node: nodeToInsert)
      }
    }

    func removeNodeBindings(node: Node) {
      node.previous?.next = node.next
      node.next?.previous = node.previous
      node.next = nil
      node.previous = nil
    }
  }
}

File: linked-list-construction/LinkedListConstruction.java
import java.util.*;

class LinkedListConstruction {
  static class DoublyLinkedList {
    public Node head;
    public Node tail;

    public void setHead(Node node) {
      if(head == null) {
        head = node;
        tail = node;
        return;
      }
      insertBefore(head, node);
    }

    public void setTail(Node node) {
      if(tail == null) {
        setHead(node);
        return;
      }
      insertAfter(tail, node);
    }

    public void insertBefore(Node node, Node nodeToInsert) {
      if(nodeToInsert == head && nodeToInsert == tail) { return ; }
      remove(nodeToInsert);
      nodeToInsert.next = node;
      nodeToInsert.prev = node.prev;
      if(node.prev == null) {
        head = nodeToInsert;
      } else {
        node.prev.next = nodeToInsert;
      }
      node.prev = nodeToInsert;
    }

    public void insertAfter(Node node, Node nodeToInsert) {
      if(nodeToInsert == head && nodeToInsert == tail) { return ; }
      remove(nodeToInsert);
      nodeToInsert.prev = node;
      nodeToInsert.next = node.next;
      if(node.next == null) {
        tail = nodeToInsert;
      } else {
        node.next.prev = nodeToInsert;
      }
      node.next = nodeToInsert;
    }

    public void insertAtPosition(int position, Node nodeToInsert) {
      if(position == 1) { setHead(nodeToInsert); return ; }
      Node node = head;
      int currentPosition = 1;
      while(node != null && currentPosition != position) {
        currentPosition++;
        node = node.next;
      }

      if(node != null) {
        insertBefore(node, nodeToInsert);
      } else {
        setTail(nodeToInsert);
      }
    }

    public void removeNodesWithValue(int value) {
      Node node = head;
      while(node != null) {
        Node nodeToRemove = node;
        node = node.next;
        if(nodeToRemove.value == value) {
          remove(nodeToRemove);
        }
      }
    }

    public void remove(Node node) {
      if(node == head) {
        head = head.next;
      }
      if(node == tail) {
        tail = tail.prev;
      }
      removeNodeBindings(node);
    }

    public void removeNodeBindings(Node node) {
      if(node.prev != null) {
        node.prev.next = node.next;
      }
      if(node.next != null) {
        node.next.prev = node.prev;
      }
      node.next = null;
      node.prev = null;
    }

    public boolean containsNodeWithValue(int value) {
      Node node = head;
      while(node != null && node.value != value) {
        node = node.next;
      }
      return node != null;
    }
  }

  // Do not edit the class below.
  static class Node {
    public int value;
    public Node prev;
    public Node next;

    public Node(int value) {
      this.value = value;
    }
  }
}
